mws/mws.py

```diff
@@ -109,83 +102,119 @@ def clean_params(params):
+    apiresponse.original
+    """
+
+    def __init__(self, data, rootkey=None, force_cdata=False):
         """
-        Provides access to the parsed contents of an XML response as a tree of ObjectDicts.
+        Besides the recommended apiresponse.parsed property, we offer a couple
+        of useful attributes here.
         """
-        if self._rootkey:
-            return self._response_dict.get(self._rootkey, self._response_dict)
-        return self._response_dict
+        # Fallback, raw and meta attributes for xml and textfiles
+        self.original = data  # requests.request response object, link above
+        self.headers = self.original.headers  # just easier to access
 
+        # Recommended attributes only for xml
+        self.pydict = None  # alternative to xml parsed or dot_dict
+        self.dot_dict = None  # fallback for xml parsed
 
-class DataWrapper(object):
-    """
-    Text wrapper in charge of validating the hash sent by Amazon.
-    """
+        # Recommended attribute only for textdata
+        self.textdata = None
 
-    def __init__(self, data, headers):
-        self.original = data
-        self.response = None
-        self.headers = headers
-        if 'content-md5' in self.headers:
-            hash_ = utils.calc_md5(self.original)
-            if self.headers['content-md5'].encode() != hash_:
-                raise MWSError("Wrong Content length, maybe amazon error...")
+        # parsing
+        self._rootkey = rootkey
+        self._force_cdata = force_cdata
+        self._main()
 
-    @property
-    def parsed(self):
+    def _main(self):
+        """
+        Try different parsing strategies.
+        """
+        # a better guess for the correct encoding
+        self.original.encoding = self.guess_encoding()
+        textdata = self.original.text
+        # We don't trust the amazon content marker.
+        try:  # try to parse as xml
+            self._xml2dict(textdata)
+        except ExpatError:  # if it's not xml its a plain textfile, like a csv
+            self.textdata = textdata
+
+    def guess_encoding(self):
+        """
+        Detecting encoding incrementally with a hotfix.
+        """
+        # fix for one none ascii character
+        chardet.utf8prober.UTF8Prober.ONE_CHAR_PROB = 0.26
+        bytelist = self.original.content.splitlines()
+        detector = chardet.UniversalDetector()
+        for line in bytelist:
+            detector.feed(line)
+            if detector.done:
+                break
+        detector.close()
+        return detector.result['encoding']
+
+    def _xml2dict(self, rawdata):
+        """
+        Parse XML with xmltodict to a python dictionary.
         """
-        Similar to the `parsed` property of DictWrapper, this provides a similar interface for a data response
-        that could not be parsed as XML.
+        namespaces = self._extract_namespaces(rawdata)
+        self._mydict = xmltodict.parse(rawdata, dict_constructor=dict,
```

Comment:
#121 we decided we would prefer to use the python library xmltodict in future. If you parse the xml like that you not only remove '/n' you also parse cdata correct.